File: top_interview_questions/1_easy/4_trees/validate_binary_search_tree/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.valid(root, float('-inf'), float('inf'))

    def valid(self, root, min_, max_):
        if not root:
            return True

        return min_ < root.val < max_ and self.valid(root.left, min_, root.val) and self.valid(root.right, root.val, max_)

File: top_interview_questions/1_easy/4_trees/validate_binary_search_tree/solution.go
package main

import (
	"fmt"
	"math"
)

func main() {
	tree := &TreeNode{
		Val: 2,
		Left: &TreeNode{
			Val: 1,
		},
		Right: &TreeNode{
			Val: 3,
		},
	}
	fmt.Print(isValidBST(tree)) // expected true
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return helper(root, math.MinInt16, math.MaxInt16)
}

func helper(root *TreeNode, min, max int) bool {
	if root == nil {
		return true
	}

	if root.Val < min || root.Val > max {
		return false
	}

	return helper(root.Left, min, root.Val) && helper(root.Right, root.Val, max)
}
